### ISSUE TITLE
Writing a ChunkedStream may pause until there is no more input

### ISSUE THREAD

**Issue by mfulgo** [netty/netty#3311]
**Netty version:** 3.10.0.Final

**Context:**
Chunked downloads of a large InputStream can lag writing data to the client until the whole stream is read if the socket's output buffer becomes full.

**To reproduce:**
1. Start the below application
2. Run `wget -S 'http://localhost:8080/index'`

At some point during the download, it should stop receiving data.  I've had it happen as early as 0% (<3Mb) or as late as 100MB into the transfer. If you wait long enough, the InputStream will finish being read and the download will very quickly complete, sending the remaining data in a very short amount of time.

If you remove the line `bootstrap.setOption("child.sendBufferSize", 65536);`, it may not freeze. If you add a break point at Line 222 of AbstractNioWorker, you'll catch the threads when the problem manifests.

**Example application:**
```java
package io.netty;

import org.jboss.netty.bootstrap.ServerBootstrap;
import org.jboss.netty.channel.*;
import org.jboss.netty.channel.socket.nio.NioServerSocketChannelFactory;
import org.jboss.netty.handler.codec.http.*;
import org.jboss.netty.handler.stream.ChunkedStream;
import org.jboss.netty.handler.stream.ChunkedWriteHandler;

import java.io.IOException;
import java.io.InputStream;
import java.net.InetSocketAddress;
import java.util.Random;
import java.util.concurrent.Executors;

public class Example {
    public static class VeryLargeStream extends InputStream {
        public static int LIMIT = 500000000;
        private int pos = 0;
        private Random random = new Random(1L);

        @Override
        public int read() throws IOException {
            pos++;
            if (pos % 1000000 == 0) try {
                Thread.sleep(random.nextInt(500));
            } catch (InterruptedException e) {}
            return pos <= LIMIT ? 0 : -1;
        }

        @Override
        public int available() throws IOException {
            return LIMIT - pos > 0 ? random.nextInt(LIMIT - pos) : 0;
        }
    }

    public static class ExampleHandler extends SimpleChannelUpstreamHandler {
        private int FIVE_KB_CHUNK_SIZE = 5 * 1024;

        @Override
        public void messageReceived(ChannelHandlerContext ctx, MessageEvent e) throws Exception {
            if (e.getMessage() instanceof HttpRequest) {
                DefaultHttpResponse response = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
                response.setChunked(true);
                response.headers().set(HttpHeaders.Names.TRANSFER_ENCODING, HttpHeaders.Values.CHUNKED);
                response.headers().set(HttpHeaders.Names.CONTENT_LENGTH, VeryLargeStream.LIMIT);
                e.getChannel().write(response);
                e.getChannel().write(new ChunkedStream(new VeryLargeStream(), FIVE_KB_CHUNK_SIZE));
            } else {
                super.messageReceived(ctx, e);
            }
        }
    }

    public static void main( String[] args ) {
        int port = 8080;
        String host = "0.0.0.0";

        NioServerSocketChannelFactory socketChannelFactory = new NioServerSocketChannelFactory(
                Executors.newCachedThreadPool(),
                Executors.newCachedThreadPool());

        ServerBootstrap bootstrap = new ServerBootstrap(socketChannelFactory);

        bootstrap.setPipelineFactory(new ChannelPipelineFactory() {
            @Override
            public ChannelPipeline getPipeline () {
                ChannelPipeline pipeline = Channels.pipeline();
                pipeline.addLast("HttpRequestDecoder", new HttpRequestDecoder());
                pipeline.addLast("HttpChunkAggregator", new HttpChunkAggregator(1 * 1024 * 1024));
                pipeline.addLast("HttpResponseEncoder", new HttpResponseEncoder());
                pipeline.addLast("Streamer", new ChunkedWriteHandler());
                pipeline.addLast("ExampleHandler", new ExampleHandler());

                return pipeline;
            }
        });

        bootstrap.setOption("child.sendBufferSize", 65536);

        bootstrap.bind(new InetSocketAddress(host, port ));
    }
}
```

```
$ java -version
java version "1.7.0_65"
OpenJDK Runtime Environment (IcedTea 2.5.3) (7u71-2.5.3-0ubuntu0.14.04.1)
OpenJDK 64-Bit Server VM (build 24.65-b04, mixed mode)
```

Operating system: Ubuntu Linux 14.04.1 64-bit

```
$ uname -a
Linux myhost 3.13.0-43-generic #72-Ubuntu SMP Mon Dec 8 19:35:06 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by mfulgo** [comment]
A few more details on this:
The problem doesn't occur *because* of changing the sendBufferSize, that simply encourages the problem. (I was able to reproduce it without changing the sendBufferSize, although less frequently.)

It seems to occur when the `AbstractNioWorker` isn't able to transfer a full buffer's worth of bytes into the `WriteableByteChannel`.  It drops into [this else block](https://github.com/netty/netty/blob/netty-3.10.0.Final/src/main/java/org/jboss/netty/channel/socket/nio/AbstractNioWorker.java#L222) and suspends writing to the channel.

While that seems like the right thing to do for a time, it doesn't actually restart sending data over the socket until the rest of the InputStream is read.  (Subsequent calls to `write` and eventually `sendDownStream` eventually get to [line 139 in `AbstractNioWorker`](https://github.com/netty/netty/blob/netty-3.10.0.Final/src/main/java/org/jboss/netty/channel/socket/nio/AbstractNioWorker.java#L139).) In the case of a very large file (or a slow or infinite stream), the client request may time-out because it hasn't received any data in a very long time.

**Comment by mfulgo** [comment]
I did some testing with Netty 4.0.25.Final and can confirm that this is not an issue as of that version.  `NioSocketChannel` doesn't use the `writeSuspend` flag like 3.x did, which resolves the issue.

Please feel free to close this if you have no plans to address this in the 3.x branch.

**Comment by trustin** [comment]
Thank you for your patience @mfulgo. Let me close this issue for now.

**Comment by mfulgo** [comment]
This is was also fixed in the 3.10 branch, as of version 3.10.2.Final.  I'm pretty sure it was fixed by https://github.com/netty/netty/commit/452df0454d34fdd5e0022f40c20ad7c45e7f1895

**Comment by trustin** [comment]
That's great. Thanks for following-up, @mfulgo
